oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+
+As a result, the Open edX "micro-frontend" architecture focused on creating an unopinionated, loosely coupled set of applications with the goal of enabling teams to iterate quickly and independently. This goal was successful.
+
+However, now roughly 6 years into this gradual re-platforming, we've discovered that the completely siloed nature of these MFEs has created its own set of problems.  Our frontends are more like independent single-page applications than micro-frontends, as we never invested in ways of composing them together. The problems inherent in today's MFE architecture are described below.
+
+Consistency
+===========
+
+Because each MFE is completely independent, this can create inconsistencies from MFE to MFE.
+
+Reusable components (Paragon)
+-----------------------------
+
+MFEs may use different versions of ``@openedx/paragon``, resulting in functional and stylistic differences.
+
+Header/Footer
+-------------
+
+MFEs may use different versions of ``@edx/frontend-component-header`` and ``@edx/frontend-component-footer``, also resulting in functional, stylistic, and content/navigation differences.
+
+Branding
+--------
+
+Brand packages created from ``@openedx/brand-openedx`` may be different versions, resulting in any number of subtle visual differences.
+
+Other Dependencies
+------------------
+
+MFEs may have completely different versions of any other dependency. We mitigate some of this by consolidating some important dependencies in ``@openedx/frontend-build`` and ``@edx/frontend-platform``, but even those can have different versions from MFE to MFE. For developers, this increases cognitive load and slows velocity because of the need to adjust to the idiosyncracies of each application.
+
+User and Developer Experience
+=============================
+
+Bundle Size
+-----------
+
+Each MFE bundles all of its own dependencies, even if they're the same version as another MFE. This means that as a user navigates between MFEs, they end up downloading the same set of common dependencies over and over again. This can result in megabytes of overhead for the average user navigating between a few MFEs, and slows down the entire experience.
+
+Full Page Refreshes
+-------------------
+
+Each MFE has its own index.html page, and needs to load all its own dependencies from scratch whenever you navigate to it. This means that the browser performs a full page refresh each time a user transfers from MFE to MFE.
+
+Build time
+----------
+
+Each MFE must bundle all its own dependencies at build time, creating significant overhead for the Webpack build process. When building multiple MFEs, this additional, repeated overhead adds up quickly, making builds prohibitively slow.
```

Comment:
Are you discussing developer time, operator/release time, or both? I'm reading this with an eye to establishing which personas currently experience difficulties in the current regime

Reply:
Both developers and operators are affected by this, as any change to anything in an MFE requires a Webpack rebuild.  But while developers only have to deal with this one MFE at a time - presumably the one they're developing at a given moment - operators have it worse: whenever they want a change in a specific MFE, they have to rebuild all of them.  (At least the way Tutor does things now.)

Reply:
Great, that's what I thought.

```suggestion
Each MFE must bundle all its own dependencies at build time, creating significant overhead for the Webpack build process. When building multiple MFEs, this additional, repeated overhead adds up quickly, making builds for operators as well as developers prohibitively slow.
```